### PR TITLE
Runtime Banner for Mobile Deep Linking

### PR DESCRIPTION
**Pull request by fdoxyz**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Implements [RFC 72](https://github.com/forem/rfcs/pull/72)

## Related Tickets & Documents

https://github.com/forem/forem-ios/pull/47

## QA Instructions, Screenshots, Recordings

1. Boot app locally
2. Enable the runtime_banner Feature Flag
   - I like to use the console for this so Flipper.enable(:runtime_banner) would do the trick
3. Open any article, user page, homepage, etc on your **desktop browser**, **Android browser** or **Forem/DEV iOS app**
   - You shouldn't be able to see a banner with the text "Open with Forem app"
4. Open any page using your **iOS device**
   - You should be able to see a banner with the text "Open with Forem app"
5. Click on the banner and you should be able to see the deep linking flow

This is what it looks like right now **when the app is installed**:

https://user-images.githubusercontent.com/6045239/115251430-c8b31080-a0e7-11eb-9132-4161bd0bd4e7.MP4

This is what it looks like **when the app isn't installed**:

https://user-images.githubusercontent.com/6045239/113349663-a18dce00-92f5-11eb-8c98-5d07d2725308.MP4

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://docs.forem.com/frontend/accessibility)._

Yes, I want the banner to be accessible but haven't made sure it is yet. Any suggestions to improve here are very welcome.

## Added tests?

- [x] Yes

## [Forem core team only] How will this change be communicated?

- [x] This change does not need to be communicated, and this is why not: This feature will only go live the day we release version `1.0` of the Forem iOS app to the AppStore, that's the reason it's behind a `FeatureFlag`.

## [optional] Are there any post deployment tasks we need to perform?

None

## [optional] What gif best describes this PR or how it makes you feel?

This is me trying to make the banner to look "okay":

css family guy

### CODE REVIEW

File: app/assets/stylesheets/runtime-banner.scss [hunk 1]
```diff
@@ -1,15 +1,74 @@
+@import 'variables';
+@import '_mixins';
+
 .runtime-banner {
```
Comment (fdoxyz) [inline app/assets/stylesheets/runtime-banner.scss:4]: This entire `.scss` file can be improved/refactored. This is just what I was able to put together for it to "work" 😅

File: app/assets/javascripts/initializers/initializeRuntimeBanner.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    // the retry button in case the user will need it
+    const targetLink = `com.forem.app://${window.location.host}${targetPath}`;
+    retryButton.href = targetLink;
+    launchCustomSchemeDeepLink(targetLink);
+  } else if (Runtime.currentOS() === 'Android') {
+    const targetIntent =
+      'intent://scan/#Intent;scheme=zxing;package=com.google.zxing.client.android;end';
+    retryButton.href = targetIntent;
+    installNowButton.href =
+      'https://play.google.com/store/apps/details?id=to.dev.dev_android';
```
Comment (fdoxyz) [inline app/assets/javascripts/initializers/initializeRuntimeBanner.js:60]: The Android side is a placeholder. Still a WIP

File: app/assets/javascripts/initializers/initializeRuntimeBanner.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+// A separate function is needed for this because it's impossible to test a
+// custom scheme launch in Cypress. More about this here:
+// https://medium.com/cypress-io-thailand/understand-stub-spy-apis-that-will-make-your-test-better-than-ever-on-cypress-io-797cb9eb205a
+function launchCustomSchemeDeepLink(targetURL) {
+  window.location.href = targetURL;
+}
+// Making it visible for testing (in order to stub it)
+window.launchCustomSchemeDeepLink = launchCustomSchemeDeepLink;
```
Comment (fdoxyz) [inline app/assets/javascripts/initializers/initializeRuntimeBanner.js:17]: This is the best I was able to do in order to test it (check out the `stub` in the Cypress test)

File: app/controllers/deep_links_controller.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+            appIDs: forem_app_ids,
+            components: [{ "/" => "*" }]
+          },
+        ]
+      },
+      webcredentials: {
+        apps: forem_app_ids
+      }
+    }
+  end
```
Comment (fdoxyz) [inline app/controllers/deep_links_controller.rb:20]: Instead of serving a static file this will allow us to customize the AASA file

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -104,15 +104,21 @@
       </div>
     <% end %>
     <%= yield %>
+    <% if FeatureFlag.enabled?(:runtime_banner) || Rails.env.test? %>
```
Comment (fdoxyz) [inline app/views/layouts/application.html.erb:107]: Added the `Rails.env.text?` bit so the banner will be visible in E2E tests. This entire condition will be removed when the `FeatureFlag` is gone

**Review by fdoxyz — COMMENTED** [review]
This is a working version but it needs help/input from @forem/design. The RFC was [intentionally left open to flesh out the design details](https://github.com/forem/rfcs/pull/72#pullrequestreview-596733490). This is a summary of the logic behind the deep linking for iOS devices:

> When a user taps on the Deep Link banner they will be sent to `/r/mobile`, which is a pivot page. This page will immediately send users into the mobile app if possible. When users don't have the app installed they will be prompted with 3 actions: "Go back", "Try again" or "Install the app". The last one will open the AppStore directly (since we haven't released we have DEV iOS link as placeholder).

Support for Android devices is a WIP. It's not critical right now because the banner will only be visible to iOS users.

This is what it looks like right now **when the app is installed**:

https://user-images.githubusercontent.com/6045239/113349642-9b97ed00-92f5-11eb-820b-0df8269a874d.MP4

This is what it looks like **when the app isn't installed**:

https://user-images.githubusercontent.com/6045239/113349663-a18dce00-92f5-11eb-8c98-5d07d2725308.MP4

**Comment by ludwiczakpawel** [comment]
@fdoxyz any thoughts about these? https://www.figma.com/file/TXwxPdXJgiOdcVjuBahhVI/Universal-link-to-open-mobile-app?node-id=0%3A1 i remember sharing these in some RFC but can't really find where it was :P

**Comment by fdoxyz** [comment]
> any thoughts about these? https://www.figma.com/file/TXwxPdXJgiOdcVjuBahhVI/Universal-link-to-open-mobile-app?node-id=0%3A1 i remember sharing these in some RFC but can't really find where it was :P

Absolutely! I did embed those four options directly in the RFC (linked in the description of the PR). This is just my best attempt at implementing option 4 so far 😳

I was trying option 4 because it was the one I liked the most at first. After trying it out locally I'm starting to think Option 1 or 2 would be a little less intrusive and still give us the benefit of presenting the banner though.

I'll be sure to mention this during tomorrow's Design crit 👍

**Comment by ludwiczakpawel** [comment]
@fdoxyz gotcha! awesome :) i'm happy to help you out with implementing frontend for the banner :)

**Comment by fdoxyz** [comment]
Just an FYI: I have an idea that could enhance the flow. 

If it works I'll push the changes here, but this is the gist of it:

image

**Comment by fdoxyz** [comment]
Hi @ludwiczakpawel, it's probably worth merging this PR with the FeatureFlag despite the Frontend not being fleshed out yet. We really need to test this on a production environment before launching. Let me know what you think.

File: app/assets/javascripts/initializers/initializeRuntimeBanner.js (hunk 3 above)
Comment (nickytonline) [reply app/assets/javascripts/initializers/initializeRuntimeBanner.js:17]: It's global by default, so adding it to the window shouldn't be necessary.

e.g. here's some function globally accessible from the base.js

image

File: app/assets/javascripts/initializers/initializeRuntimeBanner.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
+/* global Runtime */
+
+function handleDismissRuntimeBanner(event) {
```
Comment (nickytonline) [inline app/assets/javascripts/initializers/initializeRuntimeBanner.js:3]: These could also live/get imported into base.jsx then they are in an ES module not polluting the global.

File: cypress/integration/loggedOutFlows/runtimeBannerDeepLinking.spec.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  });
+
+  it('should redirect to custom scheme when someone taps on the banner', () => {
+    const deepLinkPath = '/custom_path';
+    const launchCustomSchemeDeepLinkStub = (url) => {
+      expect(url).to.eq(`com.forem.app:/${deepLinkPath}`);
+    };
+    cy.window()
+      .then((win) => {
+        cy.stub(
```
Comment (nickytonline) [inline cypress/integration/loggedOutFlows/runtimeBannerDeepLinking.spec.js:39]: Is stubbing this required or can we use [cy,intercet](https://docs.cypress.io/api/commands/intercept#Usage) instead and handle when the URL is `com.forem.app:/${deepLinkPath}` is hit?

File: app/assets/javascripts/initializers/initializeRuntimeBanner.js (hunk 3 above)
Comment (fdoxyz) [reply app/assets/javascripts/initializers/initializeRuntimeBanner.js:17]: Just tested it locally and you're totally right, tests continue to pass without this so it's unnecessary. I'm removing the window assignment here 👍

File: cypress/integration/loggedOutFlows/runtimeBannerDeepLinking.spec.js (hunk 7 above)
Comment (fdoxyz) [reply cypress/integration/loggedOutFlows/runtimeBannerDeepLinking.spec.js:39]: I think this is required because when redirecting to custom schemes (like `tel://`) are never accessed (Cypress/browser don't allow it).

[I read about it here](https://medium.com/cypress-io-thailand/understand-stub-spy-apis-that-will-make-your-test-better-than-ever-on-cypress-io-797cb9eb205a#b6ad) and some other sites a few weeks ago and this is why I went with this approach.

I'm not sure if I'm missing something but I just gave it another try using intercept and wasn't able to make it work.

File: app/assets/javascripts/initializers/initializeRuntimeBanner.js (hunk 6 above)
Comment (fdoxyz) [reply app/assets/javascripts/initializers/initializeRuntimeBanner.js:3]: By making them "live/get imported into base.jsx" do you mean they would be added to `window.Forem.handleDismissRuntimeBanner`? Or just placing them inside that file without a namespace (like [`initializeNav()`](https://github.com/forem/forem/blob/main/app/javascript/packs/base.jsx#L92))?

I'm thinking that in a way it makes sense to keep them close to the rest of the logic here, otherwise when revisiting this code we would have to go looking in multiple files for these functions. It's not a strong opinion that I want to defend though, just looking for confirmation on this before moving them.

Comment (nickytonline) [reply app/assets/javascripts/initializers/initializeRuntimeBanner.js:3]: You could do it like `initializeNav()` unless there is some other reason it needs to be exposed on the global object.

Comment (fdoxyz) [reply app/assets/javascripts/initializers/initializeRuntimeBanner.js:3]: I did my best attempt at moving it there but fell into an InstantClick edge case(s) and wasn't able to get out of it (some Cypress tests started failing in some cases). Keeping it here helps a lot because it's ensured to run on all new pages (including InstantClick navigations).

If we really want to move it to `base.jsx` it would be great if we can make that a separate effort to avoid blocking this PR for longer. Otherwise I'll probably need help sorting out a good way to move the logic there.

File: app/assets/javascripts/initializers/initializeRuntimeBanner.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+// A separate function is needed for this because it's impossible to test a
+// custom scheme launch in Cypress. More about this here:
+// https://medium.com/cypress-io-thailand/understand-stub-spy-apis-that-will-make-your-test-better-than-ever-on-cypress-io-797cb9eb205a
+function launchCustomSchemeDeepLink(targetURL) {
+  window.location.href = targetURL;
+}
+
+function initializeRuntimeBanner() {
+  // This will provide the dismiss functionality for the Runtime Banner
+  const bannerDismiss = document.querySelectorAll(
```
Comment (citizen428) [inline app/assets/javascripts/initializers/initializeRuntimeBanner.js:19]: Random question: if we only need the first element, can't we use `querySelector` instead of `querySelector(...)[0]`?

File: app/controllers/deep_links_controller.rb [hunk 9]
```diff
@@ -0,0 +1,21 @@
+class DeepLinksController < ApplicationController
+  def mobile; end
+
+  # Apple Application Site Association
+  def aasa
+    supported_apps = ["R9SWHSQNV8.com.forem.app"]
```
Comment (citizen428) [inline app/controllers/deep_links_controller.rb:6]: I'm not overly familiar with deep-linking, is hard-coding this here a good idea? Should it maybe be some config value instead? Or an env var or something?

File: app/views/layouts/application.html.erb [hunk 10]
```diff
@@ -104,15 +104,21 @@
       </div>
     <% end %>
     <%= yield %>
+    <% if FeatureFlag.enabled?(:runtime_banner) || ENV['E2E'] == 'true' %>
+      <div class="Browser-iOS">
+        <div class="runtime-banner">
+          <a href="https://udl.visualcosita.com/?r=<%= CGI.escape(URL.url("/r/mobile?deep_link=#{request.path}")) %>" class="flex items-center flex-1">
```
Comment (citizen428) [inline app/views/layouts/application.html.erb:110]: Is this your domain? Will we eventually move the app behind it to a more "official" one?

File: app/controllers/deep_links_controller.rb (hunk 9 above)
Comment (fdoxyz) [reply app/controllers/deep_links_controller.rb:6]: Hardcoding this here allows for the Forem app to be supported by default. Actually all forems have the DEV app hardcoded right now ([https://community.codenewbie.org/apple-app-site-association](https://community.codenewbie.org/apple-app-site-association) for example). Now that's definitely not ideal, because the DEV app only supports DEV compared to the Forem app that will support any Forem.

Ideally I'm hoping to continue to expand on exposing configurations creators will be able to leverage in order to have control over this. Adding a selector to `/admin/consumer_apps` where they can choose which one of their consumer apps will be the one users will be deep linked into from the banner shouldn't be too difficult, but felt like I wanted to test this out with the TestFlight builds before going too deep in the weeds on configurability. Check out the [**feature 3** section of the RFC](https://github.com/forem/rfcs/blob/d28ba6ce61c2cda9d4018b4237ec4866be4359b9/text/0072-runtime-banner-with-deep-linking.md#admin-panel-configuration-feature-3) for more details.

tl;dr: It's not bad (specially compared to the current state we're on) and adding some configuration to make this more dynamic is indeed going to happen (soon 🤞 )

File: app/views/layouts/application.html.erb (hunk 10 above)
Comment (fdoxyz) [reply app/views/layouts/application.html.erb:110]: Yes, this is my domain. I wanted to ship this (still behind the FeatureFlag) so it's tangible and validated (make sure it's tested with the TestFlight build) while we wait for a final call and execution of the "official one".

File: app/assets/javascripts/initializers/initializeRuntimeBanner.js (hunk 6 above)
Comment (fdoxyz) [reply app/assets/javascripts/initializers/initializeRuntimeBanner.js:3]: @nickytonline please let me know if this sounds alright with you whenever you have a chance. Thanks!

Comment (nickytonline) [reply app/assets/javascripts/initializers/initializeRuntimeBanner.js:3]: Sounds good @fdoxyz. We can defer moving things to base.jsx for now.

**Comment by fdoxyz** [comment]
Travis build needed a restart, but we're green again

File: app/assets/javascripts/initializers/initializeRuntimeBanner.js [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+function handleDismissRuntimeBanner(event) {
+  const runtimeBanner = document.querySelector('.runtime-banner');
+  if (runtimeBanner) {
+    runtimeBanner.remove();
+  }
+}
+
+// A separate function is needed for this because it's impossible to test a
+// custom scheme launch in Cypress. More about this here:
+// https://medium.com/cypress-io-thailand/understand-stub-spy-apis-that-will-make-your-test-better-than-ever-on-cypress-io-797cb9eb205a
```
Comment (vaidehijoshi) [inline app/assets/javascripts/initializers/initializeRuntimeBanner.js:12]: Small request: could we use JSDoc formatting for the functions in this file? It might make it easier for us to parse through and understand what's happening here, and hopefully that'll make it simpler to change this code in the future 👍🏽

File: app/assets/javascripts/initializers/initializeRuntimeBanner.js [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+  const targetPath = urlParams.get('deep_link');
+
+  if (Runtime.currentOS() === 'iOS') {
+    // The install now must target Apple's AppStore
+    installNowButton.href =
+      'https://apps.apple.com/us/app/dev-community/id1439094790';
+
+    // We try to deep link directly by launching a custom scheme and populate
+    // the retry button in case the user will need it
+    const targetLink = `com.forem.app://${window.location.host}${targetPath}`;
```
Comment (vaidehijoshi) [inline app/assets/javascripts/initializers/initializeRuntimeBanner.js:48]: I wonder if we could pull some of these hard-coded values into constants? My thought with that suggestion is that it would make it easier to change in the future

Just riffing, maybe something like:
```javascript
const FOREM_URL = "com.forem.app"
const ANDROID_URL = "https://play.google.com/store/apps/details?id=to.dev.dev_android"
```

File: app/views/deep_links/mobile.html.erb [hunk 13]
```diff
@@ -0,0 +1,42 @@
+<div id="mobile-deep-link-spinner" class="flex flex-col">
+  <svg class="crayons-icon crayons-spinner" width="24" height="24" viewBox="0 0 24 24" aria-hidden="true" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M18.364 5.636L16.95 7.05A7 7 0 1019 12h2a9 9 0 11-2.636-6.364z" fill="currentColor" class="grow-1"></path></svg>
+  <p class="align-center">Attempting to open the mobile app...</p>
```
Comment (vaidehijoshi) [inline app/views/deep_links/mobile.html.erb:3]:
```suggestion
  <p class="align-center">Opening the mobile app...</p>
```

Just a copy suggestion; "attempting to" makes it sound a little like it's likely to fail LOL 😂 🙊

File: app/views/deep_links/mobile.html.erb [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+</div>
+
+<div class="mobile-deep-link-banner invisible">
+  <div class="crayons-notice crayons-notice--warning">
+    <p>
+      Whoops! Are you stuck here unable to open the mobile app?
```
Comment (vaidehijoshi) [inline app/views/deep_links/mobile.html.erb:9]:
```suggestion
      Whoops! Did you get stuck trying to open the mobile app?
```

File: app/views/deep_links/mobile.html.erb [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+      </a>
+    </div>
+  </div>
+</div>
+
+<script type="text/javascript">
+  setTimeout(function() {
+    document.getElementById('mobile-deep-link-spinner').remove();
+    document.querySelectorAll('.mobile-deep-link-banner')[0].classList.remove('invisible');
+  }, 3333);
```
Comment (vaidehijoshi) [inline app/views/deep_links/mobile.html.erb:41]: I'm curious, why did we go with `3333` ms here? 🤔

File: app/views/layouts/application.html.erb (hunk 10 above)
Comment (vaidehijoshi) [reply app/views/layouts/application.html.erb:110]: Non-blocking: should we extract this url into a url helper (maybe in the `URL` module? 🤔)

File: cypress/integration/loggedOutFlows/runtimeBannerDeepLinking.spec.js [hunk 16]
```diff
@@ -0,0 +1,76 @@
+describe('Runtime Banner Deep Linking', () => {
+  const runtimeStub = {
+    onBeforeLoad: (win) => {
+      Object.defineProperty(win.navigator, 'userAgent', {
```
Comment (vaidehijoshi) [inline cypress/integration/loggedOutFlows/runtimeBannerDeepLinking.spec.js:4]: `Object.defineProperty` in the wild!!? Well there's something you don't see everyday... 👀

File: app/views/deep_links/mobile.html.erb (hunk 15 above)
Comment (fdoxyz) [reply app/views/deep_links/mobile.html.erb:41]: TBH there's no specific reason whatsoever behind this decision (100% arbitrary number)

> "Whats would be a good amount of time to let users deep link into the apps before presenting the fallback page?"

My answer to that question was "about 3 seconds, maybe" so I went with `3333` because I like the number 3 and because we never get to specify milliseconds (always do `1000/3000/5000/etc`) 😆

I can change it back to `3000` or any other number really (I have no strong opinions on this). Though it's maybe worth testing it out on our devices and see how that feels "in real life" and see if this is alright? 😄

File: app/views/layouts/application.html.erb (hunk 10 above)
Comment (fdoxyz) [reply app/views/layouts/application.html.erb]: I think it would be good to keep that logic outside the views and it also makes it easier to test there. Done ✅

Comment (fdoxyz) [reply app/views/layouts/application.html.erb]: Update: Our "official" server is being used now 😄

File: app/views/deep_links/mobile.html.erb (hunk 15 above)
Comment (vaidehijoshi) [reply app/views/deep_links/mobile.html.erb:41]: >  I like the number 3 and because we never get to specify milliseconds (always do 1000/3000/5000/etc)

haha, same! that's why i asked --  i never see anything beyond like 1000 or 5000 ms, and i was like "oh that's weird"

We could extract that and leave a comment that we picked that number because it's the right amount of time to let users deep link into apps before a fallback (that would make it evident that we picked it for a reason!); without that context, it feels a bit "magic numbery" to me :)

File: app/assets/javascripts/initializers/initializeRuntimeBanner.js [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+  // The target path comes in via GET request query params
+  const urlParams = new URLSearchParams(window.location.search);
+  const targetPath = urlParams.get('deep_link');
+
+  // Constants - they will become dynamic (configurable by creators) in upcoming releases
+  const FOREM_IOS_SCHEME = 'com.forem.app';
+  const FOREM_APP_STORE_URL =
+    'https://apps.apple.com/us/app/dev-community/id1439094790';
+  const FOREM_GOOGLE_PLAY_URL =
+    'https://play.google.com/store/apps/details?id=to.dev.dev_android';
```
Comment (vaidehijoshi) [inline app/assets/javascripts/initializers/initializeRuntimeBanner.js:61]: Love this improvement 👏🏽

File: app/lib/url.rb [hunk 18]
```diff
@@ -67,6 +67,16 @@ def self.local_image(image_name, host: nil)
...
   end
 
+  # Creates a deep link URL (for mobile) to a page in the current Forem and it
+  # relies on a UDL server to bounce back mobile users to the local `/r/mobile`
+  # fallback page. More details here: https://github.com/forem/udl-server
+  #
+  # @param path [String] the target path to deep link
+  def self.deep_link(path)
+    target_path = CGI.escape(url("/r/mobile?deep_link=#{path}"))
+    "https://forem-udl-server.herokuapp.com/?r=#{target_path}"
```
Comment (vaidehijoshi) [inline app/lib/url.rb:77]: Non-blocking comment: We could extract `https://forem-udl-server.herokuapp.com` into a constant in this file in case we end up wanting to re-use it or reference it elsewhere (but if we don't think we'll need to do that, then we can probably punt this)

File: app/controllers/deep_links_controller.rb (hunk 9 above)
Comment (vaidehijoshi) [reply app/controllers/deep_links_controller.rb:6]: I agree with @citizen428 that we should avoid hardcoding this and the below `"R9SWHSQNV8.to.dev.ios"` string -- but I'm also okay to revisit this in a separate, follow-up PR and extract it out if that seems easier (we could leave a `TODO: ` here for that if that would be helpful!)

**Review by vaidehijoshi — APPROVED** [review]

File: app/controllers/deep_links_controller.rb (hunk 9 above)
Comment (fdoxyz) [reply app/controllers/deep_links_controller.rb:8]: > we could leave a TODO: here for that if that would be helpful!

I think it would, and I just did 😃

File: app/views/deep_links/mobile.html.erb [hunk 19]
```diff
@@ -0,0 +1,48 @@
+<div id="mobile-deep-link-spinner" class="flex flex-col">
```
Comment (nickytonline) [inline app/views/deep_links/mobile.html.erb:1]: We'll need to make this markup accessible and handle it as well in the JS that shows/hides the different sections of markup.

We'll need to use [aria-live](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions). This will allow a screen reader to announce if there is a failure. We can use it in conjunction with the [aria-hidden](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-hidden_attribute) attribute (toggling between true/false).

I'm done for today, but just pinging @aitchiss, when she starts her day, to drop her thoughts here as well. 👋🏻

If you search the codeabse for `aria-live`, you'll see some examples of this in the wild.

**Review by nickytonline — CHANGES_REQUESTED** [review]
This is looking good @fdoxyz. There are some accessibility issues that need to be addressed.

File: app/views/deep_links/mobile.html.erb (hunk 19 above)
Comment (fdoxyz) [reply app/views/deep_links/mobile.html.erb:1]: Thank you for the note and suggestions @nickytonline!

From what I understand at first glance, we'd need to add `aria-live="polite"` to both divs ([this one](https://github.com/forem/forem/blob/953df231a6cf76512f1a415aaf70dcb63c4c4516/app/views/deep_links/mobile.html.erb#L1) and [this one](https://github.com/forem/forem/blob/953df231a6cf76512f1a415aaf70dcb63c4c4516/app/views/deep_links/mobile.html.erb#L6)) so they're announced.

On top of that, toggle `aria-hidden="true/false"` in the JS [here](https://github.com/forem/forem/blob/953df231a6cf76512f1a415aaf70dcb63c4c4516/app/views/deep_links/mobile.html.erb#L46) so they're available/hidden to the assistive technology when they become available/not-available.

I can make these changes (please correct me if I'm wrong though) tomorrow and also include any other suggestions @aitchiss may have 🙂

File: app/controllers/deep_links_controller.rb (hunk 9 above)
Comment (citizen428) [reply app/controllers/deep_links_controller.rb:8]: Thanks for the context @fdoxyz, a comment is good enough for now.

File: app/views/layouts/application.html.erb (hunk 10 above)
Comment (citizen428) [reply app/views/layouts/application.html.erb]: Great, thanks @fdoxyz!

**Review by citizen428 — APPROVED** [review]
Thanks for the updates @fdoxyz, apart from the a11y concerns I think this is ready to go.

**Review by rhymes — APPROVED** [review]
I tried it manually and it works perfectly.

Two small details:

- the banner looks a bit small:

IMG_5017

- the timeout to show the "oops" message, is probably a bit short, depending on how quickly the person notices the system alert to open the app, but we can tune that if we need to.

Great job!

File: app/views/deep_links/mobile.html.erb (hunk 19 above)
Comment (aitchiss) [reply app/views/deep_links/mobile.html.erb:1]: the `invisible` class you're using will hide an item from assistive technology too - so you won't need to use `aria-hidden` in addition to it 🙂 

I would suggest `aria-live="polite"` for the `Opening the mobile app...` div, but `role="alert"` for the other ('alert' is generally used for errors which feels appropriate in this case!)

File: app/views/layouts/application.html.erb [hunk 20]
```diff
@@ -104,15 +104,21 @@
...
     <%= yield %>
+    <% if FeatureFlag.enabled?(:runtime_banner) || ENV['E2E'] == 'true' %>
+      <div class="Browser-iOS">
+        <div class="runtime-banner">
+          <a href="<%= URL.deep_link(request.path) %>" target="_blank" class="flex items-center flex-1">
+            <%= inline_svg_tag("logo-forem-app.svg", aria: true, class: "crayons-icon crayons-icon--default", width: 32, height: 32, title: "Forem App icon") %>
+            <span class="ml-2">Open with the Forem app...</span>
+          </a>
+          <button type="button" class="runtime-banner__dismiss crayons-btn crayons-btn--ghost crayons-btn--icon crayons-btn--inverted crayons-btn--s pl-2">
+            <%= inline_svg_tag("x.svg", aria: true, class: "crayons-icon", title: "Dismiss the banner") %>
```
Comment (aitchiss) [inline app/views/layouts/application.html.erb:115]:
```suggestion
            <%= inline_svg_tag("x.svg", aria: true, class: "crayons-icon", title: "Dismiss banner: Open with the Forem app") %>
```

suggestion (not-blocking) - just a thought but given screen reader users will often bring up a list of all buttons/links on a page, it might help if it was more descriptive about which banner this relates to

File: app/views/layouts/application.html.erb [hunk 21]
```diff
@@ -104,15 +104,21 @@
       </div>
     <% end %>
     <%= yield %>
+    <% if FeatureFlag.enabled?(:runtime_banner) || ENV['E2E'] == 'true' %>
+      <div class="Browser-iOS">
+        <div class="runtime-banner">
+          <a href="<%= URL.deep_link(request.path) %>" target="_blank" class="flex items-center flex-1">
+            <%= inline_svg_tag("logo-forem-app.svg", aria: true, class: "crayons-icon crayons-icon--default", width: 32, height: 32, title: "Forem App icon") %>
```
Comment (aitchiss) [inline app/views/layouts/application.html.erb:111]:
```suggestion
            <%= inline_svg_tag("logo-forem-app.svg", aria_hidden: true, class: "crayons-icon crayons-icon--default", width: 32, height: 32) %>
```

nitpick (blocking) - I feel like this icon is purely decorative, in which case we don't want to surface it to screen reader users (otherwise the link is read out as "Forem App icon, Open with the Forem app...")

**Review by aitchiss — CHANGES_REQUESTED** [review]
Just a couple of very small suggestions, in addition to the `aria-live` Nick suggested 😄

File: app/assets/javascripts/initializers/initializeRuntimeBanner.js (hunk 17 above)
Comment (fdoxyz) [reply app/assets/javascripts/initializers/initializeRuntimeBanner.js:61]: I agree 100% with you @rhymes, being able to configure these is definitely part of the plan. I shared some of that context [in this comment](https://github.com/forem/forem/pull/13190#discussion_r618335855). The tl;dr is that it's contemplated by the [feature 3](https://github.com/forem/rfcs/blob/d28ba6ce61c2cda9d4018b4237ec4866be4359b9/text/0072-runtime-banner-with-deep-linking.md#admin-panel-configuration-feature-3) section of the RFC. Coming soon after we are able to confirm all the moving parts here work properly 🤓 

The implementation details will definitely require some thought, as you mentioned these are hardcoded in the JS and those configs live in the server, but I'm optimistic we'll be able to figure out an elegant solution when tackling that challenge.

File: app/views/deep_links/mobile.html.erb (hunk 19 above)
Comment (fdoxyz) [reply app/views/deep_links/mobile.html.erb]: Cool, the `role="alert"` vs `aria-live="polite"` is great insight to keep in mind for the future. Thanks a lot @aitchiss!

**Review by aitchiss — APPROVED** [review]
It would be great if someone with an iOS device could double check the aria announcements with VoiceOver on (unless you know a way I can do it in the simulator? - I couldn't find one), but it all looks good to me 👍

File: app/lib/url.rb (hunk 18 above)
Comment (joshpuetz) [reply app/lib/url.rb:77]: We're already playing around with this address, so I definitely agree we should extract this value!

**Comment by fdoxyz** [comment]
I sped up the voiceover so it wouldn't make for a long video, but seems to be working for me 🎉

@nickytonline since you requested changes I'm unable to merge until I get your approval. Let me know if there's anything else you'd want to validate here 🙂

https://user-images.githubusercontent.com/6045239/115894673-c23dd500-a416-11eb-93e3-7ca6769cc672.mov

**Comment by nickytonline** [comment]
> It would be great if someone with an iOS device could double check the aria announcements with VoiceOver on (unless you know a way I can do it in the simulator? - I couldn't find one), but it all looks good to me 👍 

I'll test this @aitchiss.

**Review by nickytonline — APPROVED** [review]
Tested this locally on iOS and works great. As mentioned, in Slack, we should take a peek at Edge on iOS and Firefox as well.

**Review by nickytonline — APPROVED** [review]
Tested this locally on iOS and works great. As mentioned, in Slack, we should take a peek at Edge on iOS and Firefox as well. 🚀

**Comment by rhymes** [comment]
@nickytonline [my test](https://github.com/forem/forem/pull/13190#pullrequestreview-643018464) was with Firefox on iOS so we're covered there. After all... they're all versions of Safari :D
